### ISSUE TITLE
oao yarn invocation errors often give insufficient information

### ISSUE THREAD

**Issue by edmorley** [guigrpa/oao#52]
In #51, the `yarn link neutrino` command failed with:

```
2017-10-10T14:09:45.076Z                 main TRACE ┌── Run cmd yarn link neutrino at packages/neutrino-middleware-env [CREATED]
2017-10-10T14:10:03.867Z           storyboard WARN  └── REVEALED PAST LOGS END HERE
2017-10-10T14:10:03.867Z                 yarn ERROR | Killed
2017-10-10T14:10:03.868Z                 yarn INFO  | 
2017-10-10T14:10:04.008Z                 main ERROR Command 'yarn link neutrino' failed at packages/neutrino-middleware-env
                                         main ERROR   name: 'Error'
                                         main ERROR   message: Command failed: /bin/sh -c yarn link neutrino
                                         main ERROR   message: Killed
                                         main ERROR   message: 
                                         main ERROR   stack: Error: Command failed: /bin/sh -c yarn link neutrino
                                         main ERROR   stack: Killed
                                         main ERROR   stack: 
                                         main ERROR   stack:     at Promise.all.then.arr (/home/travis/build/edmorley/neutrino-dev/node_modules/oao/node_modules/execa/index.js:210:11)
                                         main ERROR   stack:     at process._tickCallback (internal/process/next_tick.js:109:7)
2017-10-10T14:10:07.704Z                 main TRACE └── Run cmd yarn link neutrino at packages/neutrino-middleware-env [CLOSED]
Error: Command failed: yarn link neutrino
    at _callee2$ (/home/travis/build/edmorley/neutrino-dev/node_modules/oao/lib/utils/shell.js:173:19)
    at tryCatch (/home/travis/build/edmorley/neutrino-dev/node_modules/regenerator-runtime/runtime.js:65:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/home/travis/build/edmorley/neutrino-dev/node_modules/regenerator-runtime/runtime.js:303:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as throw] (/home/travis/build/edmorley/neutrino-dev/node_modules/regenerator-runtime/runtime.js:117:21)
    at step (/home/travis/build/edmorley/neutrino-dev/node_modules/oao/lib/utils/shell.js:28:191)
    at /home/travis/build/edmorley/neutrino-dev/node_modules/oao/lib/utils/shell.js:28:402
    at process._tickCallback (internal/process/next_tick.js:109:7)
2017-10-10T14:10:07.712Z           storyboard INFO  └── ROOT STORY: Node.js 6.11.4 on Linux 64-bit, SB 3.1.3 [CLOSED]
```

This error message doesn't really help with debugging the problem, since:
* What does "Killed" mean? Did the yarn command hang and the oxo exec call have to kill it? If so ideally there would be output explaining this.
* Where is the stdout/stderr from yarn? Even if the command timed out, there should still be the header row from yarn, unless output buffering is causing issues?
* If yarn didn't time out, but just gave an unhelpful stdout/stderr, then perhaps the exit code should be shown? (Maybe only if `>1` ?)
* In what directory was the command being run? The log says "Command ... failed at packages/neutrino-middleware-env", however I presume that's not the real directory, given how yarn link is supposed to work? Knowing how to run the command exactly as oao is, will help with reproducing.
* There's quite a bit of duplication between the "main ERROR" prefixed lines and the non-storyboard prefixed lines later.

Many thanks :-)

**Comment by guigrpa** [comment]
Hmmmm, I'm afraid oao is not generating the `Killed` message. It is indeed the `stderr` output: in the following line, notice the `yarn` bit (source of the message: either `stdout` or `stderr` of the command being launched) and the `ERROR` bit (hence, `stderr`):

```
2017-10-10T14:10:03.867Z                 yarn ERROR | Killed
```

Wrt. other comments:

* The missing yarn header: maybe yarn gets killed (by whom?) before printing the header out?
* I agree we should print out the exit code.
* I don't understand your 4th bullet. In principle, it *does* run `yarn link neutrino` from the `packages/neutrino-middleware-env` directory (in the first bootstrap phase, it had already run `yarn link` from the `packages/neutrino` directory). Maybe I'm missing something?
* I agree with the unnecessary duplication, but it would have a lower priority right now…

**Comment by edmorley** [comment]
Ah perhaps it's an OOM? 
https://stackoverflow.com/questions/10876136/what-can-cause-node-js-to-print-killed-and-exit

I believe we could confirm that by seeing if the exit code was 137 (http://www.tldp.org/LDP/abs/html/exitcodes.html).

> The missing yarn header: maybe yarn gets killed (by whom?) before printing the header out?

If it's an OOM I would have thought it would occur after the point where yarn has output the version header. As such I would guess either:
* the output is being buffered (which is not necessarily a bad thing, but perhaps the buffering can be made to flush more often?)
* `execa` (or oao) is ignoring the stdout when there's something on `stderr` to display (I would find this unlikely)

> I don't understand your 4th bullet. In principle, it does run yarn link neutrino from the packages/neutrino-middleware-env directory (in the first bootstrap phase, it had already run yarn link from the packages/neutrino directory). Maybe I'm missing something?

Ah I was thinking of the first stage, I don't fully understand what oao does after that.

> I agree with the unnecessary duplication, but it would have a lower priority right now…

Agreed :-)

**Comment by guigrpa** [comment]
Patched in v0.10.4 to log error code upon command failure.

I couldn't find traces of buffered `stderr` output. In fact, in my examples I got `stdout` and `stderr` interspersed, as could be expected in some cases. Example:

```
2017-10-10T20:45:06.934Z                 main INFO  ┌── Run cmd yarn add dfjdfdifdjfdijfdijfdoifjdiojfoi at packages/neutrino [CREATED]
2017-10-10T20:45:07.243Z                 yarn INFO  | yarn add v1.1.0
2017-10-10T20:45:07.384Z                 yarn INFO  | [1/4] Resolving packages...
2017-10-10T20:45:09.441Z                 yarn ERROR | error Couldn't find package "dfjdfdifdjfdijfdijfdoifjdiojfoi" on the "npm" registry.
2017-10-10T20:45:09.441Z                 yarn INFO  | info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
2017-10-10T20:45:09.451Z                 yarn INFO  |
2017-10-10T20:45:09.454Z                 main ERROR Command 'yarn add dfjdfdifdjfdijfdijfdoifjdiojfoi' failed [1] at packages/neutrino
2017-10-10T20:45:09.456Z                 main INFO  └── Run cmd yarn add dfjdfdifdjfdijfdijfdoifjdiojfoi at packages/neutrino [CLOSED]
2017-10-10T20:45:09.458Z           storyboard INFO  └── ROOT STORY: Node.js 8.6.0 on Darwin 64-bit, SB 3.1.3 [CLOSED]
```

**Comment by edmorley** [comment]
Many thanks :-)
